api/src/main/java/org/eclipse/microprofile/lra/annotation/Compensate.java

```diff
@@ -27,42 +27,131 @@
+ * Therefore combining this annotation with an <code>&#64;LRA</code> annotation that does not
+ * start a new LRA will result in a <code>412 PreCondition Failed</code> JAX-RS response
+ * code. On the other hand, combining it with an <code>&#64;LRA</code> annotation that
+ * begins a new LRA can in certain uses case make sense, but in this case the LRA
  * that this method is being asked to compensate for will be unavailable.
+ * </p>
+ *
+ * <p>
+ * If the method is a JAX-RS resource method (or is a non JAX-RS method
+ * annotated with <code>&#64;Compensate</code> with return type
+ * <code>javax.ws.rs.core.Response)</code> then the following are the only
+ * valid response codes:
+ * </p>
+ *
+ *<table border="0" cellpadding="3" cellspacing="0"
+ *   summary="Valid JAX-RS compensation response codes">
+ * <caption>
+ *     <span>JAX-RS Compensation Response Codes</span>
+ *     <span>&nbsp;</span>
+ * </caption>
+ * <tr>
+ *   <th scope="col">Code</th>
+ *   <th scope="col">Response Body</th>
+ *   <th scope="col">Meaning</th>
+ * </tr>
+ * <tr>
+ *   <td scope="row">200</td>
+ *   <td scope="row">Empty</td>
+ *   <td scope="row">The resource has successfully compensated</td>
+ * </tr>
+ * <tr>
+ *   <td scope="row">202</td>
+ *   <td scope="row">Empty</td>
+ *   <td scope="row">The resource is still attempting compensation</td>
+ * </tr>
+ * <tr>
+ *   <td scope="row">404</td>
+ *   <td scope="row">Empty</td>
+ *   <td scope="row">The resource does not know about the LRA</td>
+ * </tr>
+ * <tr>
+ *   <td scope="row">412</td>
+ *   <td scope="row">Empty</td>
+ *   <td scope="row">The resource cannot join the LRA because the LRA is no
+ *   longer active</td>
+ * </tr>
+ * <tr>
+ *   <td scope="row">500</td>
+ *   <td scope="row">{@link ParticipantStatus} enum value</td>
+ *   <td scope="row"><p>The resource has failed to compensate.
+ *   The payload contains the reason for the failure.
+ *   A participant MUST remember this state until its
+ *   {@link Forget} method is called.</p>
+ *   <p>The actual value is not important but it MUST
+ *   correspond to a valid {@link ParticipantStatus} enum value. For example,
+ *   if compensation was not possible because the resource already
+ *   completed (without being asked to) then a value such as
+ *   {@link ParticipantStatus#Completed} would be appropriate or
+ *   if it was due to a generic failure then
+ *   {@link ParticipantStatus#FailedToCompensate} would be valid.
+ *   </p>
+ *   <p>
+ *   Note that the
+ *   actual state as reported by the {@link Status} method MUST
+ *   be {@link ParticipantStatus#FailedToCompensate}</p></td>
+ * </tr>
+ * </table>
+ *
+ * <p>
+ * If any other code is returned (or, in the 500 case, the body does not
+ * correspond to a valid state) then the implementation MUST either keep
+ * retrying or attempt to discover the status by calling the
+ * {@link Status} method if present or a combination of both.
```

Comment:
when the retry for `@Compensate` and calls to `@Status` method can be combined? The status method is saying that when it's present then `@Compensate` cannot be invoked multiple times or maybe I am missing something. I would suggest removing "or a combination of both".

Reply:
The reason I added "or a combination of both" is to allow the following behaviour:

1. the coordinator receives an invalid JAX-RS status code from the compensate notification
2. the coordinator sees that it can use the status method to ask for the state of the compensator
3. the status method reports that the compensator is still in the `Active` state
4. the coordinator reissues the compensate notification

ie the coordinator has resolved the ambiguity by using a combination of both strategies. Hence the bit of extra text is required.

Reply:
Very interesting point indeed. There are two ways of how we can look at this. I will open a new issue for 1.x, if you don't mind.

Reply:
I see any need for an issue

Reply:
There is other text in the spec that covers this

Reply:
I may be confused, can you respond to #187, please? We can resolve this conversation.

Reply:
Section [3.2.4. Compensating Activities](https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc#324-compensating-activities) for example includes the text:

> If there is no @Status then the @Compensate or @Complete methods will continue to be invoked until the implementation knows it has the final status.

There are other places where I talk about things being retried "... knows it has the final status".

Reply:
The scenario described here https://github.com/eclipse/microprofile-lra/pull/176/files#r302883356 is only possible when the first Compensate request never reached the participant. Otherwise the call to status would never be Active.

In that sense, the code of the Compensate will only be executed once when there is a Status method.

Of course, when developer mess things up, by returning Active from the Status method even after the Complete method is already called, multiple calls can happen.

I'm fine with the text "or a combination of both" but then we should maybe add the scenario as an example so that users and implementors know the background.

Reply:
@rdebusscher I made two changes:

- include the scenario at the end of the javadoc
- changed MUST keep retrying to SHOULD

This is too avoid the situation where the implementation keeps retrying forever.

Reply:
@rdebusscher I have also just updated the @Status and @Forget javadoc to use SHOULD for the same reason with some text about logging a warning if it decides to stop retrying. Notice that compensation violations are allowed in the spec provided that the system writes a log section as described in the [section 3.1](https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc#31-the-model):

> ... take recovery actions in the event of failure, but allowing always the possibility that recovery isn’t possible which must be logged or flagged for the administrator and manual intervention may be necessary to restore an application’s consistency.

Reply:
@mmusgrov Ok, thanks.